examples/jsm/objects/BatchedMesh.js

```diff
@@ -0,0 +1,461 @@
+		this._maxVertexCount = maxVertexCount;
+		this._maxIndexCount = maxIndexCount;
+
+		this._geometryInitialized = false;
+		this._currentGeometryCount = 0;
+		this._currentVertexCount = 0;
+		this._currentIndexCount = 0;
+
+		// Local matrix per geometry by using data texture
+		// @TODO: Support uniform parameter per geometry
+
+		this._matrices = []; 
+		this._matricesArray = null;
+		this._matricesTexture = null;
+		this._matricesTextureSize = null;
+
+		// @TODO: Calculate the entire binding box and make frustomCulled true
+		this.frustomCulled = false;
+
+		this._customUniforms = {
+			batchingTexture: { value: null },
+			batchingTextureSize: { value: 0 }
+		};
+
+		this._initMatricesTexture();
+		this._initShader();
+
+	}
+
+	_initMatricesTexture() {
+
+		// layout (1 matrix = 4 pixels)
+		//      RGBA RGBA RGBA RGBA (=> column1, column2, column3, column4)
+		//  with  8x8  pixel texture max   16 matrices * 4 pixels =  (8 * 8)
+		//       16x16 pixel texture max   64 matrices * 4 pixels = (16 * 16)
+		//       32x32 pixel texture max  256 matrices * 4 pixels = (32 * 32)
+		//       64x64 pixel texture max 1024 matrices * 4 pixels = (64 * 64)
+
+		let size = Math.sqrt( this._maxGeometryCount * 4 ); // 4 pixels needed for 1 matrix
+		size = MathUtils.ceilPowerOfTwo( size );
+		size = Math.max( size, 4 );
+
+		const matricesArray = new Float32Array( size * size * 4 ); // 4 floats per RGBA pixel
+		const matricesTexture = new DataTexture( matricesArray, size, size, RGBAFormat, FloatType );
+
+		this._matricesArray = matricesArray;
+		this._matricesTexture = matricesTexture;
+		this._matricesTextureSize = size;
+
+		this._customUniforms.batchingTexture.value = this._matricesTexture;
+		this._customUniforms.batchingTextureSize.value = this._matricesTextureSize;
+
+	}
+
+	_initShader() {
+
+		const material = this.material;
+		const currentOnBeforeCompile = material.onBeforeCompile;
+		const customUniforms = this._customUniforms;
+
+		material.onBeforeCompile = function onBeforeCompile ( parameters, renderer ) {
+
+			// Is this replacement stable across any materials?
+			parameters.vertexShader = parameters.vertexShader
+				.replace(
+					'#include <skinning_pars_vertex>',
+					'#include <skinning_pars_vertex>\n'
+						+ batchingParsVertex
+				)
+				.replace(
+					'#include <skinnormal_vertex>',
+					'#include <skinnormal_vertex>\n'
+						+ batchingbaseVertex
+						+ batchingnormalVertex
+				)
+				.replace(
+					'#include <skinning_vertex>',
+					'#include <skinning_vertex>\n'
+						+ batchingVertex
+				);
+
+			for ( const uniformName in customUniforms ) {
+
+				parameters.uniforms[ uniformName ] = customUniforms[ uniformName ];
+
+			}
+
+			// for debug
+			// console.log( parameters.vertexShader, parameters.uniforms );
+
+			currentOnBeforeCompile.call( this, parameters, renderer );
+
+		};
+
+		material.defines = material.defines || {};
+		material.defines.BATCHING = false;
+
+	}
+
+	applyGeometry( geometry ) {
+
```

Comment:
Earlier proposals had included a matrix argument here, like:

```javascript
addGeometry( geometry, worldMatrix )
```

Is there a reason you prefer to skip that? Or is it just something we'd wait and consider later?

I think most users will need to apply a matrix, and doing that internally might help cut down boilerplate. E.g. adding the same tree in several locations and scales, users will need to make two copies of the geometry and keep resetting and overwriting the second, it's a bit awkward.

Reply:
Oops, never mind — `.setMatrixAt( ... )` is fine!